### ISSUE TITLE
Need help with detecting traffic_cones in nuScenes dataset using PointPillars

### ISSUE THREAD

**Issue by ldtho** [traveller59/second.pytorch#439]
Hi everyone, 
I'm trying to build a NN that detect cones from nuScenes dataset using PointPillars, but the NN hasn't detected any cones. During training, it only spits out detections for the background (with is 0, and 1 is cones). I tried to tune many parameters in the config file but the result hasn't improved :(
Any recommendation would be appreciated.
Here is my config used to train the Pointpillars
`model: {
  second: {
    network_class_name: "VoxelNet"
    voxel_generator {
      full_empty_part_with_mean: false
      point_cloud_range : [-50, -20, -3 , 30, 20, 3] #[-80, -20, -2 , 60, 40, 3] # will use [0,1,3,4] means [-100,-100,100,1000]
      voxel_size : [0.1, 0.1, 3]
      max_number_of_points_per_voxel : 100
      #block_filtering: true # filter voxels by block height
      #block_factor: 1 # height calc width: voxel_size * block_factor * block_size= (0.05 * 1 * 8)
      #block_size: 8
      #height_threshold: 0.2 # locations with height < height_threshold will be removed.
    }
    voxel_feature_extractor: {
      module_class_name: "PillarFeatureNet"
      num_filters: [64]
      with_distance: false
      num_input_features: 5
    }
    middle_feature_extractor: {
      module_class_name: "PointPillarsScatter"
      downsample_factor: 1
      num_input_features: 64
    }
    rpn: {
      module_class_name: "RPNV2"
      layer_nums: [3, 5, 5]
      layer_strides: [1, 2, 2]
      num_filters: [64, 128, 256]
      upsample_strides: [1, 2, 4]
      num_upsample_filters: [128, 128, 128]
      use_groupnorm: false
      num_groups: 32
      num_input_features: 64
    }
    loss: {
      classification_loss: {
        weighted_sigmoid_focal: {
          alpha: 0.75 #0.25
          gamma: 2.0 #2.0
          anchorwise_output: true
        }
      }
      localization_loss: {
        weighted_smooth_l1: {
          sigma: 3.0
          code_weight: [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0]
        }
      }
      classification_weight: 1.0
      localization_weight: 1.0
    }
    num_point_features: 5 # model's num point feature should be independent of dataset
    # Outputs
    use_sigmoid_score: true
    encode_background_as_zeros: true #false
    encode_rad_error_by_sin: true
    sin_error_factor: 1.0

    use_direction_classifier: false # this can help for orientation benchmark
    direction_loss_weight: 0.2 # enough.
    num_direction_bins: 2
    direction_limit_offset: 1

    # Loss
    pos_class_weight: 1.0
    neg_class_weight: 1.0 #1.0

    loss_norm_type: NormByNumPositives
    # Postprocess
    post_center_limit_range: [-50, -20, -3 , 30, 20, 3] #[-120, -60, -4 , 80, 60, 5] #[0, -40, -2.2, 70.4, 40, 0.8]
    nms_class_agnostic: false # only valid in multi-class nms

    box_coder: {
      ground_box3d_coder: {
        linear_dim: false
        encode_angle_vector: false
      }
    }
    target_assigner: {
      class_settings: {
        class_name: "traffic_cone"
        anchor_generator_range: {
          sizes: [0.36, 0.36, 0.8] # wlh
          anchor_ranges: [-50, -20, -3 , 30, 20, 3]
          rotations: [0, 1.57] # DON'T modify this unless you are very familiar with my code.
          # custom_values: [0, 0] # velocity vector base value
        }
#        anchor_generator_stride: {
##          sizes: [0.6, 1.76, 1.73] # wlh
#          sizes: [0.36, 0.36, 0.8] # wlh
##          strides: [0.28, 0.28, 0.0] # if generate only 1 z_center, z_stride will be ignored
#          strides: [0.14, 0.14, 0.0] # if generate only 1 z_center, z_stride will be ignored
##          offsets: [0.14, -20.02, -1.465] # origin_offset + strides / 2
#          offsets: [0.07, -20.09, -1.465] # origin_offset + strides / 2
##          rotations: [0, 1.57] # 0, pi/2
#          rotations: [0, 1.57] # 0, pi/2
#        }
        matched_threshold : 0.5
        unmatched_threshold : 0.35
        use_rotate_nms: false
        use_multi_class_nms: false
        nms_pre_max_size: 1000
        nms_post_max_size: 300
        nms_score_threshold: 0.02
        nms_iou_threshold: 0.3
        region_similarity_calculator: {
          distance_similarity: {
            distance_norm: 1.414 # match range
            with_rotation: false
            rotation_alpha: 0.0 # rot error contribution
          }
        }
      }

      sample_positive_fraction : -1
      sample_size : 512
      assign_per_class: true
    }
  }
}

train_input_reader: {
  dataset: {
    dataset_class_name: "CustomNuscDataset"
    kitti_info_path: " "
    kitti_root_path: "/media/starlet/LdTho/data/sets/nuscenes/v1.0-trainval"
  }
  batch_size: 1
  preprocess: {
    num_workers: 3
    shuffle_points: true
    max_number_of_voxels: 70000

    groundtruth_localization_noise_std: [0.0, 0.0, 0.0]
    groundtruth_rotation_uniform_noise: [-0.0, 0.0]
    global_rotation_uniform_noise: [0.0, 0.0] #[-0.3925, 0.3925]
    global_scaling_uniform_noise: [0.95, 1.05]
    global_random_rotation_range_per_object: [0, 0] # pi/4 ~ 3pi/4
    global_translate_noise_std: [0.0, 0.0, 0.0] #[0.2, 0.2, 0.2]
    anchor_area_threshold: -1 # very slow if enable when using FHD map (1600x1200x40).
    remove_points_after_sample: true
    groundtruth_points_drop_percentage: 0.0
    groundtruth_drop_max_keep_points: 15
    remove_unknown_examples: false
    sample_importance: 1.0
    random_flip_x: true
    random_flip_y: true
    remove_environment: false
    database_sampler {
      database_info_path: ""
    }
  }

}

train_config: {
  optimizer: {
    adam_optimizer: {
      learning_rate: {
        one_cycle: {
          lr_max: 1e-3
          moms: [0.95, 0.85]
          div_factor: 10.0
          pct_start: 0.4
        }
      }
      weight_decay: 0.01
    }
    fixed_weight_decay: true
    use_moving_average: false
  }
  steps: 15000 # 120000 * 4 * 2 / 22680 ~ 40 epoch
  steps_per_eval: 500 #500 # 1238 * 5
  save_checkpoints_secs : 1800 # half hour
  save_summary_steps : 10
  enable_mixed_precision: true
  loss_scale_factor: -1
  clear_metrics_every_epoch: true
}

eval_input_reader: {
  dataset: {
    dataset_class_name: "CustomNuscTestDataset"
    kitti_info_path: " "
    kitti_root_path: "/media/starlet/LdTho/data/sets/nuscenes/v1.0-trainval"
  }
  batch_size: 1
  preprocess: {
    max_number_of_voxels: 120000
    shuffle_points: true
    num_workers: 3
    anchor_area_threshold: -1
    remove_environment: false
  }
}
`
